### PR TITLE
add s3 acl support

### PR DESCRIPTION
**Pull request by shichanglin5**
# What problem are we solving?

add s3 acl support
[## Access control list (ACL) overview](https://docs.aws.amazon.com/AmazonS3/latest/userguide/acl-overview.html)

The permissions related to acl and the apis involved are as follows:
image

# How are we solving the problem?

Added ACL related processing in APIs such as PutObject/PutObjectAcl/GetObject/GetObjectAcl/PutBucket/PutBucketAcl/ListObjects/GetBucketAcl

# How is the PR tested?

The relevant functional tests have been made through the s3 sdk

# Checks
- [ ] I have added unit tests if possible.
- [ ] I will add related wiki document changes and link to this PR after merging.

### CODE REVIEW

File: weed/s3api/s3_constants/extend_key.go [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package s3_constants
 
 const (
-	ExtAcpKey       = "Seaweed-X-Amz-Acp"
+	ExtAmzOwnerKey  = "Seaweed-X-Amz-Owner"
+	ExtAmzAclKey    = "Seaweed-X-Amz-Acl"
```
Comment (shichanglin5) [inline weed/s3api/s3_constants/extend_key.go:5]: Here, the owner and grants of acp are divided into two key storages, and the owner only stores the accountId (string). When verifying the object acl, the owner's check and the grants are checked separately. For the case that the owner check fails, There is no need to validate grants, thus avoiding unnecessary `json.Unmarshal(grants)` calls

eg `weed/s3api/s3api_acp.go:269`:
```go
	//owner access
	objectOwner := getAcpOwner(entry, *bucketMetadata.Owner.ID)
	if accountId == objectOwner {
		return s3err.ErrNone
	}

	//find in Grants
	acpGrants := getAcpGrants(entry)
	if acpGrants != nil {
		reqGrants := determineReqGrants(accountId, s3_constants.PermissionRead)
		for _, requiredGrant := range reqGrants {
			for _, grant := range acpGrants {
				if reflect.DeepEqual(requiredGrant, grant) {
					return s3err.ErrNone
				}
			}
		}
	}
```

File: weed/s3api/s3api_bucket_handlers.go [hunk 2]
```diff
@@ -238,44 +255,66 @@ func (s3a *S3ApiServer) hasAccess(r *http.Request, entry *filer_pb.Entry) bool {
...
 }
 
+// PutBucketAclHandler Put bucket ACL
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketAcl.html
+func (s3a *S3ApiServer) PutBucketAclHandler(w http.ResponseWriter, r *http.Request) {
+	bucket, _ := s3_constants.GetBucketAndObject(r)
+	glog.V(3).Infof("PutBucketAclHandler %s", bucket)
+
+	accountId := getAccountId(r)
+	errorCode, bucketMetadata := s3a.checkAccessForPutBucketAcl(accountId, bucket)
```
Comment (chrislusf) [inline weed/s3api/s3api_bucket_handlers.go]: Better switch the return variable order to `bucketMetadata, errorCode`

File: weed/s3api/chunked_reader_v4.go [hunk 3]
```diff
@@ -150,20 +150,20 @@ var errMalformedEncoding = errors.New("malformed chunked encoding")
 // newSignV4ChunkedReader returns a new s3ChunkedReader that translates the data read from r
 // out of HTTP "chunked" format before returning it.
 // The s3ChunkedReader returns io.EOF when the final 0-length chunk is read.
-func (iam *IdentityAccessManagement) newSignV4ChunkedReader(req *http.Request) (io.ReadCloser, s3err.ErrorCode) {
-	ident, seedSignature, region, seedDate, errCode := iam.calculateSeedSignature(req)
+func (iam *IdentityAccessManagement) newSignV4ChunkedReader(req *http.Request) (io.ReadCloser, s3err.ErrorCode, *Identity) {
```
Comment (chrislusf) [inline weed/s3api/chunked_reader_v4.go]: move the error code to the last return value

File: weed/s3api/s3api_bucket_handlers.go [hunk 4]
```diff
@@ -121,13 +121,25 @@ func (s3a *S3ApiServer) PutBucketHandler(w http.ResponseWriter, r *http.Request)
...
 	fn := func(entry *filer_pb.Entry) {
 		if identityId := r.Header.Get(s3_constants.AmzIdentityId); identityId != "" {
 			if entry.Extended == nil {
 				entry.Extended = make(map[string][]byte)
 			}
 			entry.Extended[s3_constants.AmzIdentityId] = []byte(identityId)
 		}
+
+		err = assembleEntryWithAcp(entry, acpOwner, acpGrants)
+		if err != 0 {
```
Comment (chrislusf) [inline weed/s3api/s3api_bucket_handlers.go]: s3err.ErrNone

File: weed/s3api/s3api_bucket_handlers.go [hunk 5]
```diff
@@ -121,13 +121,25 @@ func (s3a *S3ApiServer) PutBucketHandler(w http.ResponseWriter, r *http.Request)
...
+			glog.Warning(err)
```
Comment (chrislusf) [inline weed/s3api/s3api_bucket_handlers.go]: add some context to the warning

File: weed/s3api/s3api_object_handlers.go [hunk 6]
```diff
@@ -164,6 +177,12 @@ func (s3a *S3ApiServer) GetObjectHandler(w http.ResponseWriter, r *http.Request)
 	bucket, object := s3_constants.GetBucketAndObject(r)
 	glog.V(3).Infof("GetObjectHandler %s %s", bucket, object)
 
+	errCode := s3a.checkAccessForReadObject(r, bucket, object)
```
Comment (chrislusf) [inline weed/s3api/s3api_object_handlers.go]: It is reading the object metadata here, doubling the metadata read latency. not acceptable.

Comment (shichanglin5) [reply weed/s3api/s3api_object_handlers.go]: Object access validation is now offloaded to the file layer; 
at `weed/server/filer_server_handlers_read.go:87`:
```go
func (fs *FilerServer) GetOrHeadHandler(w http.ResponseWriter, r *http.Request) {

	path := r.URL.Path
	isForDirectory := strings.HasSuffix(path, "/")
	if isForDirectory && len(path) > 1 {
		path = path[:len(path)-1]
	}

	entry, err := fs.filer.FindEntry(context.Background(), util.FullPath(path))
	if err != nil {
		if path == "/" {
			fs.listDirectoryHandler(w, r)
			return
		}
		if err == filer_pb.ErrNotFound {
			glog.V(2).Infof("Not found %s: %v", path, err)
			stats.FilerRequestCounter.WithLabelValues(stats.ErrorReadNotFound).Inc()
			w.WriteHeader(http.StatusNotFound)
		} else {
			glog.Errorf("Internal %s: %v", path, err)
			stats.FilerRequestCounter.WithLabelValues(stats.ErrorReadInternal).Inc()
			w.WriteHeader(http.StatusInternalServerError)
		}
		return
	}

	//s3 acl offload to filer
	offloadHeaderBucketOwner := r.Header.Get(s3_constants.X_SeaweedFS_Header_Amz_Bucket_Owner_Id)
	if len(offloadHeaderBucketOwner) > 0 {
		if statusCode, ok := s3acl.CheckObjectAccessForReadObject(r, w, entry, offloadHeaderBucketOwner); !ok {
			w.WriteHeader(statusCode)
			return
		}
	}
        ......
```
to resolve package cycle references, create separate packages for acl and accounts
image

File: weed/server/filer_server_handlers_read.go [hunk 7]
```diff
@@ -4,6 +4,9 @@ import (
 	"bytes"
 	"context"
 	"fmt"
+	"github.com/seaweedfs/seaweedfs/weed/s3api/s3acl"
+
+	//"github.com/seaweedfs/seaweedfs/weed/s3api"
```
Comment (chrislusf) [inline weed/server/filer_server_handlers_read.go]: Delete

File: weed/s3api/s3api_acp.go [hunk 8]
```diff
@@ -1,28 +1,329 @@
...
+		if accountId == *acp.Owner.ID {
+			return acp, s3err.ErrNone
+		}
+
+		//find in Grants
+		if acp.Grants != nil {
+			reqGrants := s3acl.DetermineReqGrants(accountId, s3_constants.PermissionReadAcp)
+			for _, requiredGrant := range reqGrants {
+				for _, grant := range acp.Grants {
+					if reflect.DeepEqual(requiredGrant, grant) {
```
Comment (chrislusf) [inline weed/s3api/s3api_acp.go]: Try not to use reflection. Later there might be a need to customize the comparison.

Comment (shichanglin5) [reply weed/s3api/s3api_acp.go]: replaced with `weed/s3api/s3acl/acl.go:437`

**Pull request closed without merging**